src/security/core/src/dds_security_utils.c

```diff
@@ -0,0 +1,1141 @@
+}
+
+
+void
+DDS_Security_OctetSeq_copy(
+     DDS_Security_OctetSeq *dst,
+     const DDS_Security_OctetSeq *src)
+{
+    if (dst->_length > 0) {
+        DDS_Security_OctetSeq_deinit(dst);
+    }
+    dst->_length = src->_length;
+    dst->_maximum = src->_maximum;
+
+    if (src->_length) {
+        dst->_buffer = ddsrt_malloc(src->_length);
+        memcpy(dst->_buffer, src->_buffer, src->_length);
+    } else {
+        dst->_buffer = NULL;
+    }
+}
+
+DDS_Security_HandleSeq *
+DDS_Security_HandleSeq_alloc(void)
+{
+    DDS_Security_HandleSeq *seq;
+
+    seq = ddsrt_malloc(sizeof(*seq));
+    seq->_buffer = NULL;
+    seq->_length = 0;
+    seq->_maximum = 0;
+
+    return seq;
+}
+
+DDS_Security_long_long *
+DDS_Security_HandleSeq_allocbuf(
+     DDS_Security_unsigned_long length)
+{
+    DDS_Security_long_long *buffer;
+    buffer = ddsrt_malloc(length * sizeof(DDS_Security_long_long));
+    memset(buffer, 0, length * sizeof(DDS_Security_long_long));
+    return buffer;
+}
+
+void
+DDS_Security_HandleSeq_freebuf(
+     DDS_Security_HandleSeq *seq)
+{
+    if (!seq) {
+        return;
+    }
+    ddsrt_free(seq->_buffer);
+    seq->_maximum = 0;
+    seq->_length = 0;
+}
+
+void
+DDS_Security_HandleSeq_free(
+     DDS_Security_HandleSeq *seq)
+{
+    if (!seq) {
+        return;
+    }
+    DDS_Security_HandleSeq_freebuf(seq);
+    ddsrt_free(seq);
+}
+
+void
+DDS_Security_HandleSeq_deinit(
+     DDS_Security_HandleSeq *seq)
+{
+    if (!seq) {
+        return;
+    }
+    DDS_Security_HandleSeq_freebuf(seq);
+}
+
+
+
+
+void
+DDS_Security_Exception_set(
+     DDS_Security_SecurityException *ex,
+     const char *context,
+     int code,
+     int minor_code,
+     const char *fmt,
+    ...)
+{
+    int32_t ret;
+    size_t len;
+    char buf[1] = { '\0' };
+    char *str = NULL;
+    va_list args1, args2;
+
+    assert(context);
+    assert(fmt);
+    DDSRT_UNUSED_ARG( context );
+
+    if (!ex) {
+        return;
+    }
+
+    va_start(args1, fmt);
+    va_copy(args2, args1);
+
+    if ((ret = vsnprintf(buf, sizeof(buf), fmt, args1)) >= 0) {
+        len = (size_t)ret; /* +1 for null byte */
+        if ((str = ddsrt_malloc(len + 1)) == NULL) {
+            assert(false);
+        } else if ((ret = vsnprintf(str, len + 1, fmt, args2)) >= 0) {
+            assert((size_t) ret == len);
+        } else {
+            ddsrt_free(str);
+            str = NULL;
+        }
+    }
+
+    va_end(args1);
+    va_end(args2);
+
+    ex->message = str;
+    ex->code = code;
+    ex->minor_code = minor_code;
+}
+
+#if DDSI_INCLUDE_SSL
+DDS_EXPORT void
+DDS_Security_Exception_set_with_openssl_error(
+         DDS_Security_SecurityException *ex,
+         const char *context,
+         int code,
+         int minor_code,
+         const char *error_area
+        )
+{
+
+  BIO *bio;
+  char *buf = NULL;
+  char *str;
+  size_t len; /*BIO_get_mem_data requires long int */
+  assert(context);
+  assert(error_area);
+  assert(ex);
```

Comment:
Now that you added this, the null pointer check a few lines down (that caused the memory leak you just fixed) is superfluous ... Either it may never be a null pointer (in that case, the assert is correct and the check should go) or it is may be one (in which case the check is correct and the assert is not).

Reply:
Actually, it should not be null. I've removed the null check.